llama-index-core/llama_index/core/tools/function_tool.py

```diff
@@ -109,19 +117,27 @@ def async_fn(self) -> AsyncCallable:
 
     def call(self, *args: Any, **kwargs: Any) -> ToolOutput:
         """Call."""
-        tool_output = self._fn(*args, **kwargs)
+        tool_output = self._fn(*args, **kwargs)        
+        final_output_content = str(tool_output)
+        callback_output = self._run_callback(tool_output)  
+        if callback_output:
+            final_output_content += f" Callback: {callback_output}"
         return ToolOutput(
-            content=str(tool_output),
+            content=final_output_content,
             tool_name=self.metadata.name,
             raw_input={"args": args, "kwargs": kwargs},
             raw_output=tool_output,
         )
 
     async def acall(self, *args: Any, **kwargs: Any) -> ToolOutput:
-        """Call."""
-        tool_output = await self._async_fn(*args, **kwargs)
+        """Async Call."""
+        tool_output = self._fn(*args, **kwargs)        
+        final_output_content = str(tool_output)
+        callback_output = self._run_callback(tool_output)  
```

Comment:
Probably the callback should be async? Otherwise this will block the event loop (probably not ideal)

Reply:
Hmm, in our case we implemented this to allow requesting user confirmation and in this scenario the event loop remains blocked until the user confirms. If implementing it this way is not feasible, do you have any other suggestions?

Reply:
In my view, yes. We developed this feature so that when a FunctionTool is called, it can request user input that will influence the result or execution of that function.

That said, it makes sense for it not to be asynchronous. However, in our case, we use synchronous calls. If you believe it should be handled asynchronously, I can change it without any problems.

Reply:
Stopping to evaluate this point, the person who uses an asynchronous call in this case really doesn't make sense conceptually to stop the application.
So whoever uses the asynchronous callback in the asynchronous call will be in a different situation.
I will make the adjustment